storage,libroach: update MVCCIncrementalIterator to look at every updated key

pbardea opened the pull request:
This latest implementation of MVCCIncrementalIterator updates its use of
the time-bound iterator to only be used to skip over keys that were not
modified at all within the time bound limits.

Previously, the incremental iterator primarily used the time-bound
iterator and then used a non-time-bound iterator (sanity_iter) to check
that meta keys were properly read. See #43799 for discussion around why
this may not be sufficient.

This commit updates the implementation to use the time-bound iterator
more conservatively. The 2 iterators that it now uses are `iter` (a
non-time-bound iterator and `timeBoundIter`). The main iteration is now
done with `iter`. When `iter` sees a new key, the `timeBoundIter` is
incremented and if it goes further than `iter`, `iter` is `Seek`ed
forward to the meta key of the new key. This means that `iter` should
visit every version of every key that `timeBoundIter` visits.

It should be noted that if many adjacent keys were modified, the above
logic tries to avoid to do a `Seek` to the next key as a `Seek` is ~an
order of magnitude more expensive than a `Next`.

In cases where a lot of keys were modified, we expect to do ~double the
work since 2 iterators would need to seek to every key. However, the
expected use case for the MVCCIncrementalIterator is such that the
_vast_ majority of the data is excluded from the range (as should be the
case for incremental backups), since there is potentially years worth of
data that can be skipped over, versus hours worth of data to include.

Release note: None

pbardea committed:
storage: re-add MVCC incremental iterator tests

These tests exercise the Go MVCCIncrementalIterator which is no longer
used directly in CockroachDB. However, the MVCCIncrementalIterator is
used as an oracle to test the C++ implementation. Therefore, it would
still be beneficial to have these tests present as long as we plan to
keep these 2 implementations in sync.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45163)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
sorry about the delayed review

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-:-M0ZDYF_-8avQ3-2_idS:b-9r1x86)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea and @sumeerbhola)

___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z0QoY-j3AZ07OLeom:-M0Z0Qo_DgvBAr-ym1jh:b-ybjrtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L56)):*
> ```Go
> 
> 	// fields used for a workaround for a bug in the time-bound iterator
> 	// (#28358)
> ```

can you update the comment to also mention #43799
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z1DTf3By0sU2t3Ng6:-M0Z0dg2BOOAUkhAnf89:b-9jxs3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L85)):*
> ```Go
> 	var timeBoundIter Iterator
> 	if !opts.IterOptions.MinTimestampHint.IsEmpty() && !opts.IterOptions.MaxTimestampHint.IsEmpty() {
> 		// It is necessary for correctness that sanityIter be created before iter.
> ```

I don't think the ordering matters anymore.

Also, I think it would be good if there was a file level comment justifying the correctness, say some revised version of the following:

When using (t_s, t_e], say there is a version (committed or provisional) k@t where t is in that interval, that is visible to `iter`. All sstables containing k@t will be included in `timeBoundIter`. Note that there may be multiple sequence numbers for the key `k@t` at the storage layer, say `k@t#n1`, `k@t#n2`, ... where n1 > n2 ... some of which may be deleted, but the latest sequence number will be visible using `iter` (since not being visible would be a contradiction of the initial assumption). Since there is no delete across all sstables that deletes `k@t#n1`, there is no delete in the subset of sstables used by `timeBoundIter` that deletes `k@t#n1`, so the timeBoundIter will see `k@t`.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z5BbaD1Upcg369WP1:-M0Z5BbaD1Upcg369WP2:b-l9xotf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L117)):*
> ```Go
> func (i *MVCCIncrementalIterator) SeekGE(startKey MVCCKey) {
> 	if i.timeBoundIter != nil {
> 		i.timeBoundIter.SeekGE(startKey)
> ```

this needs to call `i.timeBoundIter.Valid()` before calling `i.timeBoundIter.Key()`
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z77UyAwkegX5LBMWL:-M0Z77UyAwkegX5LBMWM:b-ddr29o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L144)):*
> ```Go
> // key.
> func (i *MVCCIncrementalIterator) Next() {
> 	i.iter.Next()
> ```

the old code had `iter` as the one constrained by the time bounds, and the new code has `timeBoundIter` in that role. I think both `Next()` and `NextKey()` need to operate on `timeBoundIter` and then call `advance()`.

Given that `MVCCIncrementalIterator` is trying to work with both time-bounds and non-time bounds maybe the original structuring of using `iter` as either the time bounded one or not was the easier structure. Though I am wondering why it has to work in both cases. Can we just return a raw `engine.Iterator` from `NewMVCCIncrementalIterator()` if there are no time bounds? We could document as a code comment that the `engine.Iterator` returned from this "new" function only supports a subset of the `engine.Iterator` interface or explicitly make `MVCCIncremetalIterator` an interface. Is it the special handling of intents that prevents this? If so, it would be good to document more completely the behavioral contract of this iterator.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z7YjVEVLCCbrWL2Qe:-M0Z7YjVEVLCCbrWL2Qf:b-3ww0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L189)):*
> ```Go
> 
> // Advance
> // 	Pre: TBI and Iter are pointing to the same key, but perhaps different versions.
> ```

I don't see how this precondition is satisfied by the call sites.
I think the precondition that we desire is that TBI is at the next key we could possibly return, but we have to validate using Iter and keep moving forward if the validation fails.


<!-- Sent from Reviewable.io -->

pbardea committed:
storageccl: scan all versions of changed key in MVCCIncrementalIterator

This latest implementation of MVCCIncrementalIterator updates its use of
the time-bound iterator to only be used to skip over keys that were not
modified at all within the time bound limits.

Previously, the incremental iterator primarily used the time-bound
iterator and then used a non-time-bound iterator (sanity_iter) to check
that meta keys were properly read. See #43799 for discussion around why
this may not be sufficient.

This commit updates the implementation to use the time-bound iterator
more conservatively. The 2 iterators that it now uses are `iter` (a
non-time-bound iterator and `timeBoundIter`). The main iteration is now
done with `iter`. When `iter` sees a new key, the `timeBoundIter` is
incremented and if it goes further than `iter`, `iter` is `Seek`ed
forward to the meta key of the new key. This means that `iter` should
visit every version of every key that `timeBoundIter` visits.

It should be noted that if many adjacent keys were modified, the above
logic tries to avoid to do a `Seek` to the next key as a `Seek` is ~an
order of magnitude more expensive than a `Next`.

In cases where a lot of keys were modified, we expect to do ~double the
work since 2 iterators would need to seek to every key. However, the
expected use case for the MVCCIncrementalIterator is such that the
_vast_ majority of the data is excluded from the range (as should be the
case for incremental backups), since there is potentially years worth of
data that can be skipped over, versus hours worth of data to include.

Release note: None

pbardea reviewed (commented):
Thanks for the first look! I've fixed up some of the algorithm issues and added more documentation to explain what the iterator is doing and what guarantees it makes. A lot of this was based on your first pass, so thanks for that review.

I think that this is RFAL. I've also request @miretskiy take a look to double check my C++. :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-:-M0y0bXZ2IF7Cj-JES-U:b-fzuyd2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @sumeerbhola)

___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z0QoY-j3AZ07OLeom:-M0t_wguEyDuMQGC1PqX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L56)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you update the comment to also mention #43799
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z1DTf3By0sU2t3Ng6:-M0tc3W8DNx0zzY5hYTW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L85)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't think the ordering matters anymore.

Also, I think it would be good if there was a file level comment justifying the correctness, say some revised version of the following:

When using (t_s, t_e], say there is a version (committed or provisional) k@t where t is in that interval, that is visible to `iter`. All sstables containing k@t will be included in `timeBoundIter`. Note that there may be multiple sequence numbers for the key `k@t` at the storage layer, say `k@t#n1`, `k@t#n2`, ... where n1 > n2 ... some of which may be deleted, but the latest sequence number will be visible using `iter` (since not being visible would be a contradiction of the initial assumption). Since there is no delete across all sstables that deletes `k@t#n1`, there is no delete in the subset of sstables used by `timeBoundIter` that deletes `k@t#n1`, so the timeBoundIter will see `k@t`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z5BbaD1Upcg369WP1:-M0tc8VY-AVgB0msfd2N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L117)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this needs to call `i.timeBoundIter.Valid()` before calling `i.timeBoundIter.Key()`
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z77UyAwkegX5LBMWL:-M0xwQ7c7iZo__xL2IZD:bveld63) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L144)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

the old code had `iter` as the one constrained by the time bounds, and the new code has `timeBoundIter` in that role. I think both `Next()` and `NextKey()` need to operate on `timeBoundIter` and then call `advance()`.

Given that `MVCCIncrementalIterator` is trying to work with both time-bounds and non-time bounds maybe the original structuring of using `iter` as either the time bounded one or not was the easier structure. Though I am wondering why it has to work in both cases. Can we just return a raw `engine.Iterator` from `NewMVCCIncrementalIterator()` if there are no time bounds? We could document as a code comment that the `engine.Iterator` returned from this "new" function only supports a subset of the `engine.Iterator` interface or explicitly make `MVCCIncremetalIterator` an interface. Is it the special handling of intents that prevents this? If so, it would be good to document more completely the behavioral contract of this iterator.
</blockquote></details>

So, I believe that only `iter` is needed to be explicitly incremented before the advance function. The time bound iterator, as I attempted to explain in the updated comments, is only updated in `maybeSkipKeys`. In particular, I expect that `iter` be incremented by either `Next` or `NextKey` exactly once between invocations of `maybeSkipKeys`. `iter` will either be incremented in `Next` or `NextKey`, or at the end of `advance` in the case that we end up on a key outside the desired time bounds. The `timeBoundIter` is only ever advanced by `NextKey` (and so only visits one version of any key) when `iter` gets ahead of it.

`MVCCIncrementalIterator` behaves similarly to an `engine.Iterator` with the exception that the MVCCIncrementalIterator will return an error if it encounters an inline value or an intent within the given time bounds. I've added comments to note that these errors are returned from MVCCIncrementalIterator.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0Z7YjVEVLCCbrWL2Qe:-M0xwJW26csdUjFJld45:b-b320qy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf489275fa8746dd660f233f0b3012008f544ef3/pkg/storage/engine/mvcc_incremental_iterator.go#L189)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't see how this precondition is satisfied by the call sites.
I think the precondition that we desire is that TBI is at the next key we could possibly return, but we have to validate using Iter and keep moving forward if the validation fails.
</blockquote></details>

Yes, this precondition wasn't quite correct and was left over from working through the initial implementation.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

___
*[c-deps/libroach/incremental_iterator.h, line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xud7wD9zBjcg8wswj:-M0xud7wD9zBjcg8wswk:b-h2lg4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.h#L32)):*
> ```
>  large amount of keys
> ```
large number of keys?
___
*[c-deps/libroach/incremental_iterator.cc, line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xv6JI3s7cVlUlWF67:-M0xv6JI3s7cVlUlWF68:b-ym166a) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L28)):*
> ```cpp
>       write_intent(write_intent) {
> 
>   time_bound_iter.reset(nullptr);
> ```

Not needed.  If you prefer to be explicit, feel free to initialize inside argument list..

Or, if you want to be explicit, initialize it above (before "engine(engine)") or set
time_bound_iter = nullptr right in the header file.
___
*[c-deps/libroach/incremental_iterator.cc, line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xwCWr1tozXBG67RJK:-M0xwCWr1tozXBG67RJL:b-rphq00) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L56)):*
> ```cpp
> }
> 
> // getKey returns the
> ```

we would never know what getKey returns... :)
___
*[c-deps/libroach/incremental_iterator.cc, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xxBX_7YvMcE5CT3fk:-M0xxBX_7YvMcE5CT3fl:b-41t6kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L82)):*
> ```cpp
> // possibly different timestamp.
> void DBIncrementalIterator::maybeSkipKeys() {
>   if (time_bound_iter != nullptr) {
> ```

I would do 
if (tbi == nullptr) return

and de-indent the rest of the code.
___
*[c-deps/libroach/incremental_iterator.cc, line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xwmFZ8AsYRJCMKHrc:-M0xwmFZ8AsYRJCMKHrd:bkvd1sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L86)):*
> ```cpp
>     // keys.
>     std::string tbi_key = getKey(time_bound_iter->rep->key());
>     std::string iter_key = getKey(iter->rep->key());
> ```

It appears that getKey operates only on the rocksdb slices returned by the underlying rocks iterators.  As such, I think it makes sense to have getKey helper return a slice instead of making
string keys.  Besides the issues of unnecessary copies, strings are really bad at handling data
which might contain \0 characters (so I'm not even sure about the correctness of the code using strings instead of slices)
___
*[c-deps/libroach/incremental_iterator.cc, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xzB0n1WRnRHZUPqPW:-M0xzB0n1WRnRHZUPqPX:b-1aiajq) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L92)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>       DBIterNext(time_bound_iter.get(), true /* skip_current_key_versions */);
>       if (!valid) {
>         return;
>       }
> ```
</details>


This doesn't look right to me

We call (our) DBIterNext helper, which happens to return a value which we ignore.
Then we check for !valid.  Clearly, the underlying iterator might have some issues, but those are never reflected in the "valid" field of *this* incremental iterator.

It's kinda confusing, particularly, since we check for time_bound_iter->rep->Valid below.

What's also a bit confusing is that on one hand we call our helper (DBIterNext), but on the other hand we peak inside internal representation of the underlying iterator to see if the operation succeeded.

I think we should do one thing or the other... but not both.

I think we should do something like:

auto res = DBIterNext(...)
if (!res.valid) {
   // handle error
   status = res.status
   valid = false
   return
}
___
*[c-deps/libroach/incremental_iterator.cc, line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y-swv6t-nQ28HLa9a:-M0y-swv6t-nQ28HLa9b:brp6o7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L112)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>       	// same as the main iterator, we may be able to skip over a large group
>       	// of keys. The main iterator is seeked to the TBI in hopes that many
>       	// keys were skipped. Note that a Seek is an order of magnitude more
>       	// expensive than a Next call.
> ```
</details>

tabs?
___
*[c-deps/libroach/incremental_iterator.cc, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y24cH1BHNlWGNMM9H:-M0y24cH1BHNlWGNMM9I:bt9ylcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L116)):*
> ```
> !iter_rep->status().ok()) {
> ```
any benefit to this vs
iter_rep->Valid() check?
___
*[c-deps/libroach/incremental_iterator.cc, line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y335n11wIyYxYszFZ:-M0y335n11wIyYxYszF_:b-zbgnbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L224)):*
> ```cpp
> DBIterState DBIncrementalIterator::seek(DBKey key) {
>   if (time_bound_iter != nullptr) {
>     DBIterSeek(time_bound_iter.get(), key);
> ```

Same comment as above: 
  auto res = DBIterSeek(...)
  if (!res.valid) {
    // handle error
  }
___
*[c-deps/libroach/incremental_iterator.cc, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y3Fe2EqIvmkCR-6od:-M0y3Fe2EqIvmkCR-6oe:brw817a) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L233)):*
> ```cpp
>     const rocksdb::Slice iter_key(key.key.data, key.key.len);
>     if (tbi_key.compare(iter_key) > 0) {
>       key = ToDBKey(rocksdb::Slice(tbi_key));
> ```

tbi_key is already a slice -- no need to convert.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

___
*[c-deps/libroach/incremental_iterator.h, line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xud7wD9zBjcg8wswj:-M0yTz1KAX5F3af1LvOq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.h#L32)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
>  large amount of keys
> ```
large number of keys?
</blockquote></details>

Done.
___
*[c-deps/libroach/incremental_iterator.cc, line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xv6JI3s7cVlUlWF67:-M0yU9ft3M2jyOPQEDlW:b-u8ot5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L28)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Not needed.  If you prefer to be explicit, feel free to initialize inside argument list..

Or, if you want to be explicit, initialize it above (before "engine(engine)") or set
time_bound_iter = nullptr right in the header file.
</blockquote></details>

Done. Moved to header.
___
*[c-deps/libroach/incremental_iterator.cc, line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xwCWr1tozXBG67RJK:-M0yUovxA_CVv29QV2_5:bf5ky3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L56)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

we would never know what getKey returns... :)
</blockquote></details>

It's a secret. ;)
Updated.
___
*[c-deps/libroach/incremental_iterator.cc, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xxBX_7YvMcE5CT3fk:-M0yXZp4ArRTYFxWTfa8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L82)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would do 
if (tbi == nullptr) return

and de-indent the rest of the code.
</blockquote></details>

Done.
___
*[c-deps/libroach/incremental_iterator.cc, line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xwmFZ8AsYRJCMKHrc:-M0ycg5t1lREhcnKVV6S:b-u0opjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L86)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

It appears that getKey operates only on the rocksdb slices returned by the underlying rocks iterators.  As such, I think it makes sense to have getKey helper return a slice instead of making
string keys.  Besides the issues of unnecessary copies, strings are really bad at handling data
which might contain \0 characters (so I'm not even sure about the correctness of the code using strings instead of slices)

</blockquote></details>

Done, there should be no more keys as strings.
___
*[c-deps/libroach/incremental_iterator.cc, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0xzB0n1WRnRHZUPqPW:-M0yeEEMEyX5xPMQemWN:b-c2ar1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L92)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
>       DBIterNext(time_bound_iter.get(), true /* skip_current_key_versions */);
>       if (!valid) {
>         return;
>       }
> ```

This doesn't look right to me

We call (our) DBIterNext helper, which happens to return a value which we ignore.
Then we check for !valid.  Clearly, the underlying iterator might have some issues, but those are never reflected in the "valid" field of *this* incremental iterator.

It's kinda confusing, particularly, since we check for time_bound_iter->rep->Valid below.

What's also a bit confusing is that on one hand we call our helper (DBIterNext), but on the other hand we peak inside internal representation of the underlying iterator to see if the operation succeeded.

I think we should do one thing or the other... but not both.

I think we should do something like:

auto res = DBIterNext(...)
if (!res.valid) {
   // handle error
   status = res.status
   valid = false
   return
}

</blockquote></details>

Agreed. I did a pass to clean up our error checking throughout this file.
___
*[c-deps/libroach/incremental_iterator.cc, line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y-swv6t-nQ28HLa9a:-M0yeVZh8VJG6PmysgvL:b1uxe7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L112)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
>       	// same as the main iterator, we may be able to skip over a large group
>       	// of keys. The main iterator is seeked to the TBI in hopes that many
>       	// keys were skipped. Note that a Seek is an order of magnitude more
>       	// expensive than a Next call.
> ```
tabs?
</blockquote></details>

Oops, this is why one should not mix IDEs. Fixed.
___
*[c-deps/libroach/incremental_iterator.cc, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y24cH1BHNlWGNMM9H:-M0yeb540i2HnlZmauOb:bnaa18w) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L116)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> !iter_rep->status().ok()) {
> ```
any benefit to this vs
iter_rep->Valid() check?
</blockquote></details>

Updated in error checking audit.
___
*[c-deps/libroach/incremental_iterator.cc, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y3Fe2EqIvmkCR-6od:-M0ye6BB1bKXsEwwEj5P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L233)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

tbi_key is already a slice -- no need to convert.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @sumeerbhola)

___
*[c-deps/libroach/incremental_iterator.cc, line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y335n11wIyYxYszFZ:-M0ywFBf9UUBCpSXZq8O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L224)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Same comment as above: 
  auto res = DBIterSeek(...)
  if (!res.valid) {
    // handle error
  }
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

___
*[c-deps/libroach/incremental_iterator.cc, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y3Fe2EqIvmkCR-6od:-M0zAyQWCN1H8wgWeSzs:b5hjwoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L233)):*
> y is already a slice -- no need to convert.
Can you pass tbi_key directly to ToDBKey?

key = ToDBKey(tbi_key)?
___
*[c-deps/libroach/incremental_iterator.cc, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0z9h2P545-aO-RhnX0:-M0z9h2P545-aO-RhnX1:bjz9669) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/c-deps/libroach/incremental_iterator.cc#L75)):*
> ```
> rocksdb::Slice DBIncrementalIterator::getKey(rocksdb::Slice mvcc_key) {
> ```
Looking more at getKey() calls, I think we should change its signature to:

bool DBIncrementalIterator::extractKey(rocksdb::Slice mvcc_key, rocksdb::Slice *key) WARN_UNUSED_RESULT

Pass the argument "*key' down to the SplitKey, and return valid indicator.
Basically, we're making multiple calls to getKey() which mutates internal validity state;
and we never check on that. 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 106 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zGEQd-6h-pUaaxJzz:-M0zGEQe3zw6ezaSN6Qv:b-hial5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/pkg/storage/engine/mvcc_incremental_iterator.go#L106)):*
> ```Go
> 	var iter Iterator
> 	var timeBoundIter Iterator
> 	if !opts.IterOptions.MinTimestampHint.IsEmpty() && !opts.IterOptions.MaxTimestampHint.IsEmpty() {
> ```

the `MinTimestampHint` and `MaxTimestampHint` need to be consistent with `StartTime` and `EndTime`, yes? This structure is odd, since it looks like the same information is being provided twice -- can you add a code comment.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 141 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zHauN0j5jWTEqCS7p:-M0zHauN0j5jWTEqCS7q:b-eq6fg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/pkg/storage/engine/mvcc_incremental_iterator.go#L141)):*
> ```Go
> 			// Seek to the first version of the key that TBI provided.
> 			startKey = MakeMVCCMetadataKey(tbiKey)
> 		}
> ```

I am assuming the `startKey` that `SeekGE` is called with must be a metadata key, otherwise the else case would also need to construct a metadata key (since we need to notice intents). If yes, can you add a comment?
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 229 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zINXVBlD2_l6PyAla:-M0zINXVBlD2_l6PyAlb:bi3k0ns) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/pkg/storage/engine/mvcc_incremental_iterator.go#L229)):*
> ```Go
> 
> // advance advances the main iterator until it is referencing a key within
> // (start_time, end_time]. It may return an error if it encounters an
> ```

nit: "return an error" is not quite accurate since this function does not have  a return value.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 180 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zMfHR1E9uVtsADSF-:-M0zMfHR1E9uVtsADSF0:bgp4xv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L180)):*
> ```Go
> //
> // Pre: The TBI should be pointing to either the same key or the key prior
> // to the same key as the main iterator. This is enforced by a combination
> ```

Did you mean to say that the TBI is at a key <= main iterator? The phrasing "the key prior" suggests that they only differ by a one call to `NextKey()`?
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 189 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zO4n50mhIxmczD4nJ:-M0zO4n50mhIxmczD4nK:b-ttlsjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L189)):*
> ```Go
> // possibly different timestamp.
> func (i *MVCCIncrementalIterator) maybeSkipKeys() {
> 	if i.timeBoundIter != nil {
> ```

can you structure this as follows to reduce the block nesting
```
if i.timeBoundIter == nil {
  return
} 
tbiKey := ...
iterKey := ...
if iterKey.Compare(tbiKey) <= 0 {
  return
}
...
```
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 209 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zOtoa9PaZ4WPom8s2:-M0zOtobBWzM-HPkJ9l4:bwnnx1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L209)):*
> ```Go
> 			// This means that for the incremental iterator to perform a Next or
> 			// NextKey in this case, it will require only 1 extra NextKey call to the
> 			// TBI.
> ```

I think there are two things that can break this lockstep behavior:
- when the TBI is seeing phantom `MVCCKey.Key`s, i.e., keys that are completely absent from the main iterator -- this can happen rarely due to aborted transactions and keys that have been subsumed due to range tombstones.
- when the TBI has fewer keys than the the main iterator, which will be quite common.

This code doesn't handle the first case (and I think we need test coverage for it) -- it corresponds to `iterKey.Compare(tbiKey) > 0`. Since this will be rare I think we should simply call `i.timeBoundIter.SeekGE(MakeMVCCMetadataKey(iterKey))`. The code here is still correct, but just wasteful in causing the main iterator to iterate over keys that are not in the TBI.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 233 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zL6Or0Kx8uFLgoBhJ:-M0zL6Or0Kx8uFLgoBhK:b-s0f9r8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L233)):*
> ```Go
> func (i *MVCCIncrementalIterator) advance() {
> 	for {
> 		if !i.valid {
> ```

I think the invariant is that `advance()` is called when `i.valid` is true, though `i.iter.Valid()` may be true or false. In which case I don't think this if-block is needed. The only thing that will make `i.valid` false is `maybeSkipKeys()` and that is followed by an if-block that returns.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zJrvTCvzmbtyhxWxE:-M0zJrvTCvzmbtyhxWxF:b5igbc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L280)):*
> ```Go
> 			i.iter.Next()
> 			continue
> 		}
> ```

can you add a comment like the following since the use of `i.meta` is a bit confusing

// Else, i.meta contains an MVCC value and not an intent. This is handled next.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @sumeerbhola)

___
*[c-deps/libroach/incremental_iterator.cc, line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0y3Fe2EqIvmkCR-6od:-M0zfIEH2k3zUR_4dhzg:bxa9w9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/159b5d94d29badfac1c6cfed356f74b328439316/c-deps/libroach/incremental_iterator.cc#L233)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> y is already a slice -- no need to convert.
Can you pass tbi_key directly to ToDBKey?

key = ToDBKey(tbi_key)?
</blockquote></details>

Oops, should be fixed now.
___
*[c-deps/libroach/incremental_iterator.cc, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0z9h2P545-aO-RhnX0:-M0zfH6N2Zf6HWRmP0R_:b-adohok) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/c-deps/libroach/incremental_iterator.cc#L75)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> rocksdb::Slice DBIncrementalIterator::getKey(rocksdb::Slice mvcc_key) {
> ```
Looking more at getKey() calls, I think we should change its signature to:

bool DBIncrementalIterator::extractKey(rocksdb::Slice mvcc_key, rocksdb::Slice *key) WARN_UNUSED_RESULT

Pass the argument "*key' down to the SplitKey, and return valid indicator.
Basically, we're making multiple calls to getKey() which mutates internal validity state;
and we never check on that. 
</blockquote></details>

Ah, nice - updated!
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 106 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zGEQd-6h-pUaaxJzz:-M0zfRyz2kjn8BxXbdQv:b-8g67ii) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/pkg/storage/engine/mvcc_incremental_iterator.go#L106)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

the `MinTimestampHint` and `MaxTimestampHint` need to be consistent with `StartTime` and `EndTime`, yes? This structure is odd, since it looks like the same information is being provided twice -- can you add a code comment.
</blockquote></details>

I believe that the requirement is that the timestamp hint range should not be more restrictive than the range between `StartTime` and `EndTime`, though I would expect callers in practice to have these consistent.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 141 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zHauN0j5jWTEqCS7p:-M0zfwpG6yEmagdopOkC:bc60e6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/pkg/storage/engine/mvcc_incremental_iterator.go#L141)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I am assuming the `startKey` that `SeekGE` is called with must be a metadata key, otherwise the else case would also need to construct a metadata key (since we need to notice intents). If yes, can you add a comment?
</blockquote></details>

Yep, I've added a comment.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 229 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zINXVBlD2_l6PyAla:-M0zgdUw97lNJkn6updx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c618471f53c4a3a62752dec6167c641c0b2c5623/pkg/storage/engine/mvcc_incremental_iterator.go#L229)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: "return an error" is not quite accurate since this function does not have  a return value.
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 180 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zMfHR1E9uVtsADSF-:-M0zgKJ22cKCjCSHXw7k:b-hl2xza) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L180)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Did you mean to say that the TBI is at a key <= main iterator? The phrasing "the key prior" suggests that they only differ by a one call to `NextKey()`?
</blockquote></details>

Yep, updated the comment on the C++ side, but not on the go side. Updated here as well now.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 189 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zO4n50mhIxmczD4nJ:-M0zgP2406zpXQtY5Ixo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L189)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you structure this as follows to reduce the block nesting
```
if i.timeBoundIter == nil {
  return
} 
tbiKey := ...
iterKey := ...
if iterKey.Compare(tbiKey) <= 0 {
  return
}
...
```
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 209 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zOtoa9PaZ4WPom8s2:-M0zgRQyC56eBpgqC8lC:b6s9jto) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L209)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think there are two things that can break this lockstep behavior:
- when the TBI is seeing phantom `MVCCKey.Key`s, i.e., keys that are completely absent from the main iterator -- this can happen rarely due to aborted transactions and keys that have been subsumed due to range tombstones.
- when the TBI has fewer keys than the the main iterator, which will be quite common.

This code doesn't handle the first case (and I think we need test coverage for it) -- it corresponds to `iterKey.Compare(tbiKey) > 0`. Since this will be rare I think we should simply call `i.timeBoundIter.SeekGE(MakeMVCCMetadataKey(iterKey))`. The code here is still correct, but just wasteful in causing the main iterator to iterate over keys that are not in the TBI.
</blockquote></details>

Ah, I was wondering if the `iterKey > tbiKey` case was possible and didn't consider phantom keys. Thanks for pointing that out! Will follow this up with a test.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 233 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zL6Or0Kx8uFLgoBhJ:-M0zgfbB7Ivcze_jh1mq:brmm45f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L233)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think the invariant is that `advance()` is called when `i.valid` is true, though `i.iter.Valid()` may be true or false. In which case I don't think this if-block is needed. The only thing that will make `i.valid` false is `maybeSkipKeys()` and that is followed by an if-block that returns.
</blockquote></details>

@miretskiy's review of the C++ side led to a re-working of the error handling there. I've updated the go implementation to mirror how we handle errors there closer. This adds a bit more verbosity here, but I think it's more readable. I remove this check at the start of the `for` and explicitly check after every invocation which may make the iterator invalid. I think this is clearer since each error handling check is right after the code which may have caused the error.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zJrvTCvzmbtyhxWxE:-M0zhA9O8V6N2fYMw5qB:b-nyzb8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L280)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add a comment like the following since the use of `i.meta` is a bit confusing

// Else, i.meta contains an MVCC value and not an intent. This is handled next.

</blockquote></details>

Done. I was also a bit confused by the naming of `i.meta` so in addition to adding a comment here, I also added a comment to the `meta` property.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @sumeerbhola)

___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 209 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M0zOtoa9PaZ4WPom8s2:-M11zQc60oZ9gcWbU_Yl:b-j8xe4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e011e4d2fcad10931a738d90d8e51542b660892b/pkg/storage/engine/mvcc_incremental_iterator.go#L209)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Ah, I was wondering if the `iterKey > tbiKey` case was possible and didn't consider phantom keys. Thanks for pointing that out! Will follow this up with a test.
</blockquote></details>

Since it's not a correctness concern, I'd actually prefer getting this version in as soon as possible so we can start stressing it sooner. I am interested in eventually following up with a test though.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M12KOTZ1ifpdU1l70IM:-M12KOTZ1ifpdU1l70IN:b-mnh0yl) (no related file):*
Giving my LGTM for C++ -- i'll let @sumeerbhola handle go stuff.

---


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
> Reviewed 1 of 2 files at r6, 3 of 5 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 225 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M16F4xJ2SB_8fh3zkZ1:-M16F4xJ2SB_8fh3zkZ2:b-rntjc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2165527a58af138d9c5ca6ca2d59ea4dc6ebbca7/pkg/storage/engine/mvcc_incremental_iterator.go#L225)):*
> ```Go
> 		// This means that for the incremental iterator to perform a Next or NextKey
> 		// will require only 1 extra NextKey invocation while they remain in
> 		// lockstep.
> ```

btw, for performance experiments you may want to keep stats like a counter of the fast path and slow path -- unless modified keys are adjacent, the slow path will be more common.
(Not suggesting that you add that to this PR)
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 227 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M16FlCv8ExoA4us9Uui:-M16FlCv8ExoA4us9Uuj:b3ejq89) ([raw file](https://github.com/cockroachdb/cockroach/blob/2165527a58af138d9c5ca6ca2d59ea4dc6ebbca7/pkg/storage/engine/mvcc_incremental_iterator.go#L227)):*
> ```Go
> 		// lockstep.
> 
> 		if iterKey.Compare(tbiKey) > 0 {
> ```

can you restructure this as
```
cmp := iterKey.Compare(tbiKey)
if cmp > 0 {
...
}

if cmp < 0 {
...
}
```
to avoid repeated string comparisons.
___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 756 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M16QdDzBHp-kEr3vfDO:-M16QdE-4TRfQNQcy6uI:b-7utd7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2165527a58af138d9c5ca6ca2d59ea4dc6ebbca7/pkg/storage/engine/mvcc_incremental_iterator_test.go#L756)):*
> ```Go
> 	// and the second is the MVCC metadata entry (i.e. an intent). The next
> 	// SSTable contains the provisional value for the intent. The effect is that
> 	// the metadata entry is separated from the entry it is metadata for.
> ```

how is this different from `TestMVCCIncrementalIteratorIntentStraddlesSStables`?


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 225 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M16F4xJ2SB_8fh3zkZ1:-M17-VwuDzc6FYbpaHlG:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/2165527a58af138d9c5ca6ca2d59ea4dc6ebbca7/pkg/storage/engine/mvcc_incremental_iterator.go#L225)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

btw, for performance experiments you may want to keep stats like a counter of the fast path and slow path -- unless modified keys are adjacent, the slow path will be more common.
(Not suggesting that you add that to this PR)
</blockquote></details>

Ack.
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 227 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M16FlCv8ExoA4us9Uui:-M17-VQRC4tRFR5T3a8Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2165527a58af138d9c5ca6ca2d59ea4dc6ebbca7/pkg/storage/engine/mvcc_incremental_iterator.go#L227)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you restructure this as
```
cmp := iterKey.Compare(tbiKey)
if cmp > 0 {
...
}

if cmp < 0 {
...
}
```
to avoid repeated string comparisons.
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 756 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-M16QdDzBHp-kEr3vfDO:-M17-Z2O2NXdK5mAP1vD:bk1co57) ([raw file](https://github.com/cockroachdb/cockroach/blob/2165527a58af138d9c5ca6ca2d59ea4dc6ebbca7/pkg/storage/engine/mvcc_incremental_iterator_test.go#L756)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

how is this different from `TestMVCCIncrementalIteratorIntentStraddlesSStables`?
</blockquote></details>

Removed. Accidentally added when I was playing around with the idea of testing the phantom key case.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 4 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45163#-:-M17OaGm-lwuTQbBR_Tm:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

pbardea commented:
Thanks for the reviews!
bors r+
